### PR TITLE
Add fast path for Path::cmp that skips over long shared prefixes

### PR DESCRIPTION
**Pull request by the8472**
```
# before
test path::tests::bench_path_cmp_fast_path_buf_sort               ... bench:      60,811 ns/iter (+/- 865)
test path::tests::bench_path_cmp_fast_path_long                   ... bench:       6,459 ns/iter (+/- 275)
test path::tests::bench_path_cmp_fast_path_short                  ... bench:       1,777 ns/iter (+/- 34)

# after
test path::tests::bench_path_cmp_fast_path_buf_sort               ... bench:      38,140 ns/iter (+/- 211)
test path::tests::bench_path_cmp_fast_path_long                   ... bench:       1,471 ns/iter (+/- 24)
test path::tests::bench_path_cmp_fast_path_short                  ... bench:       1,106 ns/iter (+/- 9)
```

### CODE REVIEW

File: library/std/src/path.rs [hunk 1]
```diff
@@ -962,16 +962,47 @@ impl cmp::Eq for Components<'_> {}
...
+    // - backtrack to find separator before mismatch to avoid ambiguous parsings of '.' or '..' characters
+    // - if found update state to only do a component-wise comparison on the remainder,
+    //   otherwise do it on the full path
+    //
+    // The fast path isn't taken for paths with a PrefixComponent to avoid backtracking into
+    // the middle of one
+    if left.prefix.is_none() && right.prefix.is_none() {
+        // this might benefit from a [u8]::first_mismatch simd implementation, if it existed
+        match left.path.iter().zip(right.path.iter()).position(|(&a, &b)| a != b) {
+            None => return cmp::Ordering::Equal,
```
Comment (the8472) [inline library/std/src/path.rs:990]: This isn't taking length differences into account. Will fix.

File: library/std/src/path.rs [hunk 2]
```diff
@@ -962,16 +962,50 @@ impl cmp::Eq for Components<'_> {}
...
 impl cmp::Ord for Components<'_> {
     #[inline]
     fn cmp(&self, other: &Self) -> cmp::Ordering {
-        Iterator::cmp(self.clone(), other.clone())
+        compare_components(self.clone(), other.clone())
+    }
+}
+
+#[inline]
+fn compare_components(mut left: Components<'_>, mut right: Components<'_>) -> cmp::Ordering {
```
Comment (dtolnay) [inline library/std/src/path.rs:978]: This inline doesn't seem well motivated. Inlining all this into all the call sites doesn't seem necessary for the performance improvement and may even be detrimental.

```suggestion
fn compare_components(mut left: Components<'_>, mut right: Components<'_>) -> cmp::Ordering {
```

**Review by dtolnay — CHANGES_REQUESTED** [review]
Thanks! Mostly looks good -- one comment where more inline isn't necessarily better.

File: library/std/src/path.rs (hunk 2 above)
Comment (the8472) [reply library/std/src/path.rs:978]: `Path` and `Components` aren't generic, so afaik not having the `#[inline]` attribute would prevent cross-crate inlining. But I'll redo the benchmarks with and without it.

Comment (the8472) [reply library/std/src/path.rs:978]: @dtolnay microbenchmarks show no difference with or without the annotation. That said I think they're built as part of the same crate, so they would benefit from inlining that downstream users wouldn't.

See also #81281

But I can still remove it if you prefer.

Comment (the8472) [reply library/std/src/path.rs:978]: ping @dtolnay

Comment (dtolnay) [reply library/std/src/path.rs:978]: If you can't demonstrate a performance benefit to having the inline then I advise removing the inline.

This function is an enormous amount of logic (\>250 instructions). In #81281 all of those are way smaller.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: library/std/src/path.rs (hunk 2 above)
Comment (the8472) [reply library/std/src/path.rs]: Ok, removed that commit.

**Review by dtolnay — APPROVED** [review]

**Comment by Mark-Simulacrum** [comment]
This seems to have been a small [regression](https://perf.rust-lang.org/compare.html?start=7611fe438dae91084d17022e705bf64374d5ba4b&end=bcfd3f7e88084850f87b8e34b4dcb9fceb872d00&stat=instructions:u) on a number of -doc benchmarks that seems unlikely to be limited to just noise.

It might be worth trying to identify what pattern rustdoc is using in path comparisons that leads to this being slower.

@rustbot label perf-regression

**Comment by tmiasko** [comment]
Regarding regression, it looks like `OnceCell` in `imported_source_files` fall into typical `get_or_init` optimization trap, where instead of inlining the check if a cell was already initialized and outlining the init function, the init function was inlined into `get_or_init` instead, which makes the checking for initialization costly.

The init argument to `get_or_init` is often a closure, which receive an implicit inline hint, hence the reason for this behaviour. Placing an `#[inline(never)]` annotation should help. It might make sense to consider placing appropriate wrapper in the library implementation, to fix this more generally. Cases like this makes me wonder if this implicit hint for closures is actually a good idea.

**Comment by the8472** [comment]
> Cases like this makes me wonder if this implicit hint for closures is actually a good idea.

It often makes sense for iterators where the caller of the closure will also be `#[inline]`

**Comment by the8472** [comment]
The regression will be addressed by #89031

**Pull request merged**